### PR TITLE
BUG: .loc with MultiIndex with tuple in level GH#27591

### LINKED ISSUE #27591: BUG: MultiIndex Slice returns incorrect slice (compared to .xs())
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
l1 = list('abc')
l2 = [(0, 1), (1, 0)]
l3 = [0, 1]
cols = pd.MultiIndex.from_product([l1, l2, l3], names=['x', 'y', 'z'])
df = pd.DataFrame(index=range(5), columns=cols)

# This returns the expected slice
expected = df.xs((l1[0], l2[0], l3[0]), level=[0, 1, 2], axis=1)
print(expected)

# This returns an empty Series (?!)
empty1 = df.loc[:, (l1[0], l2[0], l3[0])]
# This returns an empty Series also
empty2 = df.loc[:, pd.IndexSlice[l1[0], l2[0], l3[0]]]
print(empty1)
print(empty2)

# if we omit the l2 dim on slice, we get something back
try3 = df.loc[:, pd.IndexSlice[l1[0], :, l3[0]]]
print(try3)

# Oddly enough, we can still set values on the dataframe and it works as expected.
df.loc[:, (l1[0], l2[0], l3[0])] = list(range(10, 15))
print(df)
```
#### Problem description
As far as I can understand the, all three indexing operations should provide the same result. Slicing seems to be the favored approach in the documentation, yet it doesn't return the expected result in this case.

Wrapping the tuple in a python object doesn't seem to work either, returning a KeyError.
```python
class MyTuple(object):
    def __init__(self, data):
        self.data = data
    def __eq__(self, other):
        return self.data == other.data
    def __hash__(self):
        return hash(self.data)
    def __repr__(self):
        return repr(self.data)
    def __str__(self):
        return repr(self)

l22 = [MyTuple((0,1)), MyTuple((1, 0))]
df2 = pd.DataFrame(index=range(5), columns=pd.MultiIndex.from_product([l1, l22, l3]))
df.loc[:, pd.IndexSlice[l1[0], l22[0], l3[0]]]
```

#### Expected Output
The expected result is a single series with 5 NaN values:
```python
       a                                                                                                                                                          
  (0, 1)                                                                                                                                                          
       0                                                                                                                                                          
0    NaN                                                                                                                                                          
1    NaN                                                                                                                                                          
2    NaN                                                                                                                                                          
3    NaN                                                                                                                                                          
4    NaN
```
This is the value of `expected` in the example above.

#### Output of ``pd.show_versions()``

<details>
Note that pandas 0.24.2 is being used because compatibility with Python 2 is needed.

#### INSTALLED VERSIONS
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-696.10.2.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.3.1
pip: 19.0.3
setuptools: 40.8.0
Cython: None
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.5.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2018.9
blosc: 1.7.0
bottleneck: 1.2.1
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.3.1
pymysql: 0.9.3
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #27591
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by attack68** [comment]
This PR does better than existing but its still ambiguous for pathological cases - hence why I'm floating the idea of deprecating tuples as sequence identifiers for multiple keys and relying only on lists (which can't be used as index keys).

### Effect of PR

| Before | After |
| --- | --- |
| Screen Shot 2021-07-01 at 15 26 34 | Screen Shot 2021-07-01 at 15 26 11 |

**Comment by jbrockmendel** [comment]
> hence why I'm floating the idea of deprecating tuples as sequence identifiers for multiple keys and relying only on lists (which can't be used as index keys).

I think we should do this too.  This PR is about fixing the bug in the interim.

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3232,18 +3232,37 @@ def _update_indexer(idxr: Index | None, indexer: Index | None, key) -> Index:
...
+                # GH#27591 check if this is a single tuple key in the level
+                try:
+                    lev_loc = self._get_level_indexer(k, level=i, indexer=indexer)
+                except (InvalidIndexError, TypeError, KeyError) as err:
+                    # InvalidIndexError e.g. non-hashable, fall back to treating
+                    #  this as a sequence of labels
+                    # KeyError it can be ambiguous if this is a label or sequence
+                    #  of labels
+                    #  github.com/pandas-dev/pandas/issues/39424#issuecomment-871626708
+                    for x in k:
```
Comment (jreback) [inline pandas/core/indexes/multi.py:3245]: can you put this in a method that you just call here. (the stuff in the except)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/multi.py:3245]: not nicely.  bc of namespace stuff like re-raising.  there's definitely cleanup on the horizon for these methods.

File: pandas/tests/indexing/test_loc.py [hunk 2]
```diff
@@ -2541,6 +2541,36 @@ def test_loc_with_period_index_indexer():
     tm.assert_frame_equal(df, df.loc[list(idx)])
 
 
+def test_loc_getitem_multiindex_tuple_level():
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py:2544]: does this cover the cases by @attack68 in the comment?

Comment (jbrockmendel) [reply pandas/tests/indexing/test_loc.py:2544]: only one of them i think

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel, looking forwards to cleanups on this!

**Pull request merged**